tensorflow/contrib/layers/python/layers/layers_test.py

```diff
@@ -2553,6 +2641,18 @@ def testNHWCAndNCHWTrainingProduceSameOutput(self):
       self.assertAllClose(nhwc, nchw, atol=1e-4, rtol=1e-4)
 
 
+class BatchNormFloat16Test(BatchNormTest):
+  dtype = dtypes.float16
+
+
+class BatchNormFloat32Test(BatchNormTest):
+  dtype = dtypes.float32
+
+
+class BatchNormFloat64Test(BatchNormTest):
+  dtype = dtypes.float64
+
+
 class LayerNormTest(test.TestCase):
```

Comment:
I originally updated the tests as you suggested. However, when I ran all the test cases for float16, one important test failed. I think float16 is not supported due to lack of precision? Description of the issue from #8535 reproduced below.

I do agree that 3x the cost is quite substantial, especially since most people use float32. Though compared to how long it took to build Tensorflow, the cost seem negligible, but that's not to say I advocate always increasing testing costs.

Would you think that only enabling full combinatorial testing before general releases to a more reasonable approach? Given how many changes I needed to parameterize BatchNormTest, it would likely take a good chunk of time to support optional combinatorial testing with types.

---

There seems to be an issue for supporting `float16` concerning moment calculation. Non-fused BatchNorm uses `moments()` internally and `moments()` has this caveat. Except from source code follows.

```py
    # The dynamic range of fp16 is too limited to support the collection of
    # sufficient statistics. As a workaround we simply perform the operations
    # on 32-bit floats before converting the mean and variance back to fp16
    y = math_ops.cast(x, dtypes.float32) if x.dtype == dtypes.float16 else x
```

https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/nn_impl.py#L631-L634


Running `layers_test.py` with support for `float16` and `float32` BatchNorm resulted in the following numerical errors. The `float16` moving mean values are incorrect.